File: java/patterns/src/main/java/org/prototypepatterns/creational/factorymethod/FactoryMethodPattern.java
package org.prototypepatterns.creational.factorymethod;

import java.util.List;

public final class FactoryMethodPattern {
    private FactoryMethodPattern() {
    }

    interface Button {
        String render();
    }

    static final class HtmlButton implements Button {
        @Override
        public String render() {
            return "<button>Submit</button>";
        }
    }

    static final class DesktopButton implements Button {
        @Override
        public String render() {
            return "[DesktopButton: Submit]";
        }
    }

    abstract static class Dialog {
        abstract Button createButton();

        String renderDialog() {
            return createButton().render();
        }
    }

    static final class WebDialog extends Dialog {
        @Override
        Button createButton() {
            return new HtmlButton();
        }
    }

    static final class DesktopDialog extends Dialog {
        @Override
        Button createButton() {
            return new DesktopButton();
        }
    }

    public static List<String> demo() {
        return List.of(new WebDialog().renderDialog(), new DesktopDialog().renderDialog());
    }
}

File: csharp/patterns/src/PrototypePatterns/Creational/FactoryMethod/FactoryMethodPattern.cs
namespace PrototypePatterns.Creational.FactoryMethod;

public interface IButton
{
    string Render();
}

public sealed class HtmlButton : IButton
{
    public string Render() => "<button>Submit</button>";
}

public sealed class DesktopButton : IButton
{
    public string Render() => "[DesktopButton: Submit]";
}

public abstract class Dialog
{
    protected abstract IButton CreateButton();

    public string RenderDialog() => CreateButton().Render();
}

public sealed class WebDialog : Dialog
{
    protected override IButton CreateButton() => new HtmlButton();
}

public sealed class DesktopDialog : Dialog
{
    protected override IButton CreateButton() => new DesktopButton();
}

public static class FactoryMethodPattern
{
    public static IReadOnlyList<string> Demo() => [new WebDialog().RenderDialog(), new DesktopDialog().RenderDialog()];
}
